Modules/arraymodule.c

```diff
@@ -3040,9 +3040,11 @@ array_modexec(PyObject *m)
     Py_TYPE(&PyArrayIter_Type) = &PyType_Type;
 
     Py_INCREF((PyObject *)&Arraytype);
-    PyModule_AddObject(m, "ArrayType", (PyObject *)&Arraytype);
+    if (PyModule_AddObject(m, "ArrayType", (PyObject *)&Arraytype) < 0)
```

Comment:
```suggestion
    if (PyModule_AddObject(m, "ArrayType", (PyObject *)&Arraytype) < 0) {
        Py_DECREF((PyObject *)&Arraytype);
        Py_CLEAR(m);
```

Reply:
why `Py_CLEAR(m)` ? nothing in the docs suggests we should do this... also the `binhex` module for example never clears the module pointer.

Reply:
I also think in general those references are never decremented anyways on success too, so why bother to do so in the error case?

Reply:
The references are stolen on success, but not on failure. So the failure branch has a reference leak.

Reply:
I suggested `Py_CLEAR` since the original code cleared `m` on failure. However, I'm now not entirely sure whether `Py_CLEAR`, `PY_XDECREF`, or no change to `m` at all is correct here (I see both happening in different places in the codebase).

Perhaps someone more familiar with multi-phase initialization can shed some light on this? @ncoghlan

Reply:
also the success branch has a leak, because the modules will never decref those references

Reply:
here's how the `binascii` module does it https://github.com/python/cpython/blob/master/Modules/binascii.c#L1589

it never decrefs those references on error, and it does not clear the module pointer

Reply:
Again, I don't know enough about what should be done with `m`.

But the added objects should definitely be `Py_DECREF`'ed on failure (see the note [here](https://docs.python.org/3/c-api/module.html#c.PyModule_AddObject)). `binascii` and a ton of other modules are handling this incorrectly, sure, but I don't think the bad behavior needs to spread any further!

Reply:
Good point, I now fixed the refleaks on the error paths (I think) and fixed some of the errors introduced by previous commits done by me clicking the "commit suggestion" button here on github

Reply:
`m` is a borrowed pointer to a module object managed by the caller of the `Py_mod_exec` slot, so `array_modexec` should have been leaving it alone.

That fact that it wasn't is in fact a bug dating from when the array module was first converted to multi-phase initialisation back in the original PEP 489 commit: https://github.com/python/cpython/commit/d5cacbb1d9c3edc02bf0ba01702e7c06da5bc318

Reply:
Tagging @encukou, as we both missed this back then, and it seems like an easy mistake for folks to make when converting modules from single-phase initialization to multi-phase initialization.

Reply:
Actually, lots of module init code had bad error handling. Conversion to multi-phase init is a good time to fix that up, but IMO it's an orthogonal change.
(There will hopefully be other good times to fixing this up in the future: conversion to heap types, [type slots](https://www.python.org/dev/peps/pep-0489/#possible-future-extensions).)